yaksh/views.py

```diff
@@ -121,25 +126,67 @@ def user_register(request):
     if request.method == "POST":
         form = UserRegisterForm(request.POST)
         if form.is_valid():
-            u_name, pwd, user_email, key = form.save()
-            new_user = authenticate(username=u_name, password=pwd)
-            login(request, new_user)
-            if user_email and key:
-                success, msg = send_user_mail(user_email, key)
-                context = {'activation_msg': msg}
-                return my_render_to_response(
+            if not settings.IS_DEVELOPMENT:
+                ''' Begin reCAPTCHA validation '''
+                recaptcha_response = request.POST.get('g-recaptcha-response')
+                url = 'https://www.google.com/recaptcha/api/siteverify'
+                values = {
+                    'secret': settings.GOOGLE_RECAPTCHA_SECRET_KEY,
+                    'response': recaptcha_response
+                }
+                data = urllib.urlencode(values)
+                req = urllib2.Request(url, data)
+                response = urllib2.urlopen(req)
+                result = json.load(response)
+                ''' End reCAPTCHA validation '''
+
+                if result['success']:
+                    u_name, pwd, user_email, key = form.save()
+                    new_user = authenticate(username=u_name, password=pwd)
+                    login(request, new_user)
+                    if user_email and key:
+                        success, msg = send_user_mail(user_email, key)
+                        context = {'activation_msg': msg}
+                        return my_render_to_response(
+                        request,
+                        'yaksh/activation_status.html', context
+                        )
+                else:
+                    messages.error(request, 'Invalid reCAPTCHA. Please try again.')
+                    return my_render_to_response(request,
+                            'yaksh/register.html',
+                            {'form': form,
+                            'data_key': settings.data_key
+                            }
+                            )
+                return index(request)
+            else:
+                u_name, pwd, user_email, key = form.save()
+                new_user = authenticate(username=u_name, password=pwd)
+                login(request, new_user)
+                if user_email and key:
+                    success, msg = send_user_mail(user_email, key)
+                    context = {'activation_msg': msg}
+                    return my_render_to_response(
                     request,
                     'yaksh/activation_status.html', context
-                )
-            return index(request)
+                    )
+                return index(request)
         else:
+            messages.info(request, 'Invalid form')
```

Comment:
Please add a proper message indicating what caused the invalidation of the form.